build/storage/tests/it/ptf_tests/system_tools/test_platform.py

```diff
@@ -115,46 +120,119 @@ def set_system_setup(self):
         if not self._is_docker():
             self._install_docker()
 
+    def get_pid_from_port(self, port: int):
+        return self.terminal.execute(
+            f"sudo netstat -anop | grep -Po ':{port}\s.*LISTEN.*?\K\d+(?=/)' || true"
+        )
 
-class StorageTargetPlatform(BaseTestPlatform):
+    def kill_process_from_port(self, port: int):
+        """Raise error if there is no process in specific port"""
+        pid = self.get_pid_from_port(port)
+        self.terminal.execute(f"sudo kill -9 {pid}")
 
-    def __init__(self):
-        config = StorageTargetConfig()
-        terminal = SSHTerminal(config)
-        super().__init__(terminal)
+    def clean(self):
+        self.docker.delete_running_containers()
 
-    # TODO add implementation
-    def create_subsystem(self):
-        pass
 
-    # TODO add implementation
-    def create_ramdrive(self):
-        return 'Guid'
+class StorageTargetPlatform(BaseTestPlatform):
+    def __init__(self, run_containers=True):
+        super().__init__(SSHTerminal(StorageTargetConfig()))
+        if run_containers:
+            self.docker.add_container(
+                "storage_target_container", StorageTargetContainer()
+            )
 
+    def create_subsystem(self):
+        cmd_sender_id = self.docker.get_docker_containers_id_from_docker_image_name(
```

Comment:
I changed cmd sender container as singleton

Reply:
so singleton is used to create object only once with calling the constructor multiple times. Singleton is often [treated ](https://stackoverflow.com/questions/137975/what-are-drawbacks-or-disadvantages-of-singleton-pattern)as anti-pattern. Would it be cleaner just to have a factory which would inject that created instance into all created platforms?

That factory can be beneficial later, when we need to create, for example, an instance of host target with different behavior for KVM case with vm and HW case with real host(based on config file, we will create one instance or another in the factory).

Reply:
I don't think that singleton is anti-pattern here. It can be ofc but in projects where people abuse it and trying to modify the objects during the execution. I don't see any of threats here. Singleton is just used for config classes just to make easy access of config params and not to passing this values from method to method.

Reply:
Hi Andrzej, thank you for your feedback.

You are absolutely correct that singleton can cause problems if we modify a global state inside that singleton. However, it is not the only disadvantage of it. I will refer to [this presentation](https://raw.githubusercontent.com/CppCon/CppCon2020/main/Presentations/retiring_the_singleton_pattern/retiring_the_singleton_pattern__peter_muldoon__cppcon_2020.pdf) from Peter Muldoon where he sum up all problems imposed by singleton.

- Acts as hidden dependencies in functions that use it
- No dependency injection for testing
- Initialization is out of your control
- Multiple runs can yield different results
- Usually in groups and may need initialization calls in a particular order to setup other singletons it depends on
- State is tied to program lifetime – frequently function calls in a particular order are necessary

Some of them are not applicable for us (yet?). But I'd highlight item number one. If we have a hidden dependency within a class and we should need to create more than one instance, we will have to gouge out their usage scattered around the entire codebase. So let's consider if such a situation is possible:
This thread was about cmd sender container, but you mentioned some config objects. Well... if we talk about cmd sender, potentially some customers can have weird configuration when some platforms are accessible only from any other specific platforms and they need to run multiple instances. It might be easier to change a config file and deploy the required number of container instances on the required platforms rather than changing their infrastructure. If we talk about configs, there are IPUs with multihost support where we need to have more than one instance of host-target or someone might want to create multiple storage targets to get higher performance numbers. I think, in these cases, it is better to inject required parts of a config, rather than let an object itself traverse an entire config to look for a part they represent. Especially, I do not see any problems with injecting the required objects since we have a single point(platforms factory) where we create our platforms and assemble all they need to work.